core/src/main/java/org/infinispan/container/offheap/OffHeapConcurrentMap.java

```diff
@@ -666,43 +1041,211 @@ public int size() {
+            // Normal iteration just keep adding entries until either we complete the lock bucket region or
+            // we read bytes over the read threshold
+            while (!(completedLockBucket = bucketLockStop == bucketPosition)) {
+               long address = memoryAddressHash.getMemoryAddressOffsetNoTraceIfAbsent(bucketPosition++);
+               if (address != 0) {
                   long nextAddress;
                   do {
                      nextAddress = offHeapEntryFactory.getNext(address);
-                     builder.accept(offHeapEntryFactory.fromMemory(address));
+                     values.add(offHeapEntryFactory.fromMemory(address));
+                     foundValue = true;
                   } while ((address = nextAddress) != 0);
-                  return builder.build();
-               } finally {
-                  lock.unlock();
+                  // We read a single bucket now return to get the value back
+                  break;
+               }
+            }
+            // If we completed the lock region and we haven't yet gone through the all buckets, we have to
+            // prepare for the next lock region worth of buckets
+            if (completedLockBucket && bucketPosition != bucketCount) {
+               bucketLockStop += getBucketRegionSize(bucketCount);
+            }
+         } finally {
+            readLock.unlock();
+         }
+         return foundValue;
+      }
+
+      /**
+       * Invoked when the iteration saw a bucket size less than the current bucket size of the memory lookup. This
+       * means we had a resize during iteration. In this case this method must take a slightly different approach
+       * to find the next non empty bucket. It will first determine if we can "fix" the current bucket to be an
+       * appropriate bucket in the new lookup. If this is not possible then this method will lookup the new bucket
+       * contents by searching all valid buckets that can be mapped from the old bucket. This is simple due to buckets
+       * growing in powers of two each time. Thus if a bucket position before was 4 it must map to either bucket
+       * position 8 or 9 in the new lookup (assuming only a single resize occurred).
+       * @param lockOffset the offset identifying the read lock region and lock we currently hold
+       * @param memoryAddressHash the current bucket lookup that we should read from
+       * @param newBucketSize how large the resizing bucket size is
+       * @return whether an entry was found
+       */
+      @GuardedBy("locks#readLock")
+      private boolean resizeIteration(int lockOffset, MemoryAddressHash memoryAddressHash, int newBucketSize) {
+         // If the position is divisible by the old buckets size that means we are starting a lock block region fresh
+         // thus we can just use the new memory lookup as is without doing costly iteration
+         if ((bucketPosition & (getBucketRegionSize(bucketCount) - 1)) == 0) {
```

Comment:
I think (hope) we can remove the `else` branch now that we're only iterating one bucket at once, but let's leave it for the next PR :)